src/logging.rs

```diff
@@ -213,6 +215,94 @@ fn send_memfd_payload(sock: UnixDatagram, data: &[u8]) -> Result<usize, SdError>
+            .to_str()
+            .ok_or_else(|| ValueNotUtf8Encoded(value.as_ref().to_os_string()))?;
+        let (device_s, inode_s) = s
+            .find(':')
+            .map(|i| (&s[..i], &s[i + 1..]))
+            .ok_or_else(|| ParseJournalStreamError::MissingSeparator(s.to_string()))?;
+        let device = dev_t::from_str(device_s).map_err(FailedToParseDeviceNumber)?;
+        let inode = ino_t::from_str(inode_s).map_err(FailedToParseInodeNumber)?;
+        Ok(JournalStream { device, inode })
+    }
+
+    /// Parse the device and inode number of the systemd journal stream denoted by the given environment variable.
+    ///
+    /// See [`JournalStream::parse()`] for more information.
+    pub fn from_env<S: AsRef<OsStr>>(key: S) -> Result<Self, JournalStreamEnvError> {
+        use JournalStreamEnvError::*;
+        Self::parse(
+            std::env::var_os(&key)
+                .ok_or_else(|| EnvironmentVariableUnset(key.as_ref().to_os_string()))?,
+        )
+        .map_err(|e| ParseError(key.as_ref().to_os_string(), e))
+    }
+
+    /// Parse the device and inode number of the systemd journal stream denoted by the default `$JOURNAL_STREAM` variable.
+    ///
+    /// See [`JournalStream::from_env()`] and [`JournalStream::parse()`].
+    pub fn from_env_default() -> Result<Self, JournalStreamEnvError> {
+        Self::from_env("JOURNAL_STREAM")
+    }
+
+    /// Get the journal stream that would correspond to the given file descriptor.
+    ///
+    /// Return a journal stream struct containing the device and inode number of the given file descriptor.
+    pub fn from_fd<F: AsRawFd>(fd: F) -> std::io::Result<Self> {
+        nix::sys::stat::fstat(fd.as_raw_fd())
+            .map(|stat| JournalStream {
+                device: stat.st_dev,
+                inode: stat.st_ino,
+            })
+            .map_err(std::io::Error::from)
+    }
+}
+
+/// Whether this process is directly connected to the journal.
+///
+/// Inspects the `$JOURNAL_STREAM` environment variable and compares the device and inode
+/// numbers in this variable against the stdout and stderr file descriptors.
+///
+/// Return `true` if either stream matches the device and inode numbers in `$JOURNAL_STREAM`,
+/// and `false` otherwise (or in case of any IO error).
+///
+/// Systemd sets `$JOURNAL_STREAM` to the device and inode numbers of the standard output
+/// or standard error streams of the current process if either of these streams is connected
+/// to the systemd journal.
+///
+/// Systemd explicitly recommends that services check this variable to upgrade their logging
+/// to the native systemd journal protocol.
+///
+/// See section “Environment Variables Set or Propagated by the Service Manager” in
+/// [systemd.exec(5)][1] for more information.
+///
+/// [1]: https://www.freedesktop.org/software/systemd/man/systemd.exec.html#Environment%20Variables%20Set%20or%20Propagated%20by%20the%20Service%20Manager
+pub fn connected_to_journal() -> bool {
```

Comment:
As I mentioned, this API makes little sense for consumers.

A typical logging framework will use stdout OR stderr, and will need to know if the specific one they are using (or maybe both) is attached to journald. If so, that specific stream can be upgraded to journald native protocol (e.g. for structured logging).

This detection helper doesn't really work if logging is going to stderr, but only stdout is attached to journald. We should instead separate helper for checking stdout and stderr status separately.

Reply:
We may still be missing some bits in other parts of this library for the whole thing to work, and I may be misunderstanding some pieces of this. But as far as I understand the intended usage of this would be:
 1. logging framework is journald-aware and configured to log to `stderr` (for example)
 2. on startup it checks whether the `stderr` FD is connected to the journald socket:
    * NO: log messages will be written as plaintext lines to the FD, as usual
    * YES: log messages will be sent as binary datagrams to the FD, using the [native protocol](https://systemd.io/JOURNAL_NATIVE_PROTOCOL/) to support structured logging, large blobs, etc.
 3. same applies to `stdout` FD, but it will need to be checked separately as the two streams may be configured in an asymmetric way

I haven't played with this in practice yet, so I may have misunderstood some parts of this.

Reply:
My understanding is different.  

I don't think that stdout/stderr of processes started through systemd are directly connected to the journald socket, and I don't think you can actually send journald datagrams over these streams.  Given that the journald socket is a datagram socket which uses a very specific data format how could that possibly work?  You'd also not need `$JOURNAL_STREAM` in this case: You could just `stat` the journald socket path and compare its device/inode against the result of `fstat(stdout)` and `fstat(stderr)`.

But on my system `stat /run/systemd/journal/socket` gives me a different device/inode than what I get in `$JOURNAL_STREAM`.

I guess what systemd gives you in `$JOURNAL_STREAM` is the device/inode pair of the anonymous pipe systemd created for the stdout/stderr of the process; this pipe ends somewhere inside systemd where the data gets split into lines, which get turned into native journal entries, and forwarded to the journald socket.  

Now, I haven't looked this up in the systemd sources, and I haven't found any place where this is explicitly documented, but [Stream Logging](https://www.freedesktop.org/software/systemd/man/systemd-journald.service.html#Stream%20logging) implies that there's some conversion going on, and also there's an explicit transport flag (see [TRANSPORT field](https://www.freedesktop.org/software/systemd/man/systemd.journal-fields.html#_TRANSPORT=)) for stdout which wouldn't make sense if stdout was just the journald socket, would it?

So, as I understand it, a program would check whether either stdout or stderr are connected to the journal.  In this case it'd not use stdout/stderr at all, for logging at least.  It could even close both of these FDs.

Instead it directly writes to journald, either with `sd_journal_print` and friends, or through the Unix socket API as implemented in this crate and as used by `sd_journal_print`.  It wouldn't really matter which stream the program checks, because it's not relevant which stream is connected to the journal; it's just that _some_ stream is connected to the journal, which tells you that logs likely end up in journald anyway, just unstructured, so you can just as well write to the journal yourself, in a structured way.

----

That said I can also add dedicated helpers to check stdout & stderr independently but I'd very much like to keep the simple function, because in my understanding of the whole setup it's the only function a program's likely to use.

And in any case, in the current version of this MR you could already do `JournalStream::from_env_default() == JournalStream::from_fd(std::io::stdout())` if you're really interested in checking only `stdout` :slightly_smiling_face:

Reply:
You're welcome :pray: